### PR TITLE
Merge Tape and TapedFunction

### PR DESCRIPTION
**Pull request by KDr2**

### CODE REVIEW

File: src/tapedfunction.jl [hunk 1]
```diff
@@ -1,10 +1,6 @@
 abstract type AbstractInstruction end
-
-mutable struct Tape
-    tape::Vector{<:AbstractInstruction}
-    counter::Int
-    owner
-end
+abstract type Taped end
+const Tape = Vector{AbstractInstruction}
```
Comment (devmotion) [inline src/tapedfunction.jl]: Currently it is `Vector{<:AbstractInstruction}`, is this change intended or shoild it be
```suggestion
const Tape = Vector{<:AbstractInstruction}
```

File: src/tapedfunction.jl [hunk 2]
```diff
@@ -15,50 +11,93 @@ mutable struct Instruction{F} <: AbstractInstruction
     fun::F
     input::Tuple
     output
-    tape::Tape
+    tape::Taped
```
Comment (devmotion) [inline src/tapedfunction.jl:14]: Is it supposed to be a field with a non-concrete type?

File: src/tapedfunction.jl [hunk 3]
```diff
@@ -15,50 +11,93 @@ mutable struct Instruction{F} <: AbstractInstruction
...
-Tape(owner) = Tape(Vector{AbstractInstruction}(), 1, owner)
-MacroTools.@forward Tape.tape Base.iterate, Base.length
-MacroTools.@forward Tape.tape Base.push!, Base.getindex, Base.lastindex
-const NULL_TAPE = Tape()
-
-function setowner!(tape::Tape, owner)
-    tape.owner = owner
-    return tape
+mutable struct TapedFunction <: Taped
+    func # ::Function # maybe a callable obejct
```
Comment (devmotion) [inline src/tapedfunction.jl:18]: Should this be
```suggestion
mutable struct TapedFunction{F} <: Taped
    func::F # ::Function # maybe a callable obejct
```
?

File: src/tapedfunction.jl (hunk 1 above)
Comment (KDr2) [reply src/tapedfunction.jl]: But I can't create an instance of `Vector{<:AbstractInstruction}` by `Vector{<:AbstractInstruction}()`, is there a way to deal with this?

Comment (devmotion) [reply src/tapedfunction.jl]: Maybe use a different dedicated function for whatever you want to do?

Comment (yebai) [reply src/tapedfunction.jl]: @KDr2 I suggest that we remove the immediate type `Tape` and work with `Vector{AbstractInstruction}` directly.

File: src/tapedfunction.jl [hunk 4]
```diff
@@ -77,6 +116,25 @@ function Base.show(io::IO, tp::Tape)
     print(io, String(take!(buf)))
 end
 
+## methods for Instruction
+gettape(x) = nothing
+gettape(x::Instruction) = x.tape
+function gettape(x::Tuple)
+    for i in x
+        gettape(i) != nothing && return gettape(i)
```
[the remark this answers is not in the document]
Comment (KDr2) [reply src/tapedfunction.jl]: updated.

File: src/tapedfunction.jl [hunk 5]
```diff
@@ -77,6 +116,25 @@ function Base.show(io::IO, tp::Tape)
...
+    end
+end
+
+function Base.show(io::IO, instruction::AbstractInstruction)
+    println(io, "A $(typeof(instruction))")
+end
```
[the remark this answers is not in the document]
Comment (KDr2) [reply src/tapedfunction.jl]: updated.

File: src/tapedfunction.jl (hunk 2 above)
Comment (KDr2) [reply src/tapedfunction.jl]: updated.

File: src/tapedfunction.jl [hunk 6]
```diff
@@ -1,72 +1,109 @@
...
+    println(buf, tf.ir)
+    println(buf, "------------------")
+    println(buf, "* .tape =>")
+    println(buf, "------------------")
+    println(buf, tf.tape)
+    println(buf, "------------------")
+    print(io, String(take!(buf)))
+end
+
+function run(tape::RawTape, args...)
```
Comment (yebai) [inline src/tapedfunction.jl:91]: This looks like a function that we can consider merging into `(tf::TapedFunction)(args...)`.

File: src/tapedfunction.jl [hunk 7]
```diff
@@ -77,10 +114,27 @@ function Base.show(io::IO, tp::Tape)
     print(io, String(take!(buf)))
 end
 
+## methods for Instruction
+gettape(x) = nothing
```
Comment (yebai) [inline src/tapedfunction.jl:118]: I did a quite search: it appears that `gettape` is no longer used anywhere. Perhaps consider removing it?

Comment (KDr2) [reply src/tapedfunction.jl]: Done.

File: src/tapedfunction.jl (hunk 6 above)
Comment (KDr2) [reply src/tapedfunction.jl]: done.

File: src/tapedfunction.jl [hunk 8]
```diff
@@ -1,72 +1,106 @@
...
+Base.show(io::IO, box::Box) = print(io, "Box(", box.val, ")")
 
-gettape(x) = nothing
-gettape(x::Instruction) = x.tape
-function gettape(x::Tuple)
-    for i in x
-        gettape(i) != nothing && return gettape(i)
-    end
+## methods for RawTape and Taped
+const NULL_TAPE = RawTape()
```
Comment (yebai) [inline src/tapedfunction.jl:42]: I suggest that we replace `NULL_TAPE` with `RawTape` in all places, then remove `NULL_TAPE` to further simplify the code.

File: src/tapedfunction.jl [hunk 9]
```diff
@@ -101,26 +144,9 @@ function (instr::Instruction{typeof(_new)})()
...
-        tape[1].input = input
-    end
-    for instruction in tape
-        instruction()
-        increase_counter!(tape)
-    end
-end
-
-function run_and_record!(tape::Tape, f, args...)
+function run_and_record!(tape::Taped, f, args...)
```
Comment (yebai) [inline src/tapedfunction.jl:149]: I think `run_and_record!` is a bit confusing actually, because it could mean we are running the entire `tape`. Maybe rename this funciuton to 
```suggestion
function track!(tape::Taped, f, args...)
```

File: src/tapedtask.jl [hunk 10]
```diff
@@ -66,27 +66,27 @@ TapedTask(t::TapedTask, args...) = TapedTask(func(t), args...)
 func(t::TapedTask) = t.tf.func
 
 
-function step_in(t::Tape, args)
-    len = length(t)
-    if(t.counter <= 1 && length(args) > 0)
+function step_in(tf::TapedFunction, args)
```
Comment (yebai) [inline src/tapedtask.jl:69]: Maybe consider moving `step_in` to `tapedfunction.jl`.

**Review by yebai — APPROVED** [review]
Thanks @KDr2 - it looks great. I left a few additional minor comments below. Ready to merge once they are fixed! 

It might be helpful to run the Turing tests locally before we merge this.

File: src/tapedtask.jl (hunk 10 above)
Comment (KDr2) [reply src/tapedtask.jl:69]: `step_in` involves some dealings with TapedTask, like sending produced values after an instruction is executed, so I think keeping it in `tapedtask.jl` is better.

File: src/tapedfunction.jl (hunk 9 above)
Comment (KDr2) [reply src/tapedfunction.jl]: done.

File: src/tapedfunction.jl (hunk 8 above)
Comment (KDr2) [reply src/tapedfunction.jl]: done.
